### ISSUE TITLE
Enhanced logging

### ISSUE THREAD

**Issue by enckse** [oragono/oragono#121]
Just some thoughts on logging
* It doesn't appear that any of the file logging does rolling (right?), should it? if yes, what type (date?)?
* Should probably add documentation (happy to cover this one) for what to configure on levels/settings

For the server we're running I'm likely to play with the settings so that I don't (accidentally) end up with a big oragono log file (probably enabling stdout and stderr which will go to journal on arch automagically)

**note:** I wanted to solicit thoughts but I'm not unwilling to look at making the code changes/PR but wanted to see what input others have

**Comment by enckse** [comment]
for the aur I've covered this via a logrotate rule (which only resolves this for arch/AUR users)

**Comment by euank** [comment]
I'll toss in my personal opinion:

### Log rotation

It's not the responsibility of a given program to rotate logs except in unusual circumstances. The `logrotate` tool provides a standardized and user-configurable way to configure logrotation for any copy-trunc-able files (which the log in this case should be)... this allows a sysadmin to learn a single program and configuration for log rotation rather than learning a different one per daemon installed.

A second argument against logrotation being built into the program is that as most distros move towards systemd, log-rotation is more often than not handled by the systemd-journal automatically doing it. In the case of oragono, it offers a way to turn off file logging and simply log to stdout/stderr, which is sufficient to integrate with journald and get its rotation easily.

I think that disabling the logfile in favor of stdout/stderr for a systemd setup might make more sense.

### Documentation

Definitely do think there's room for better documentation.

A sample systemd service file and oragono config associated with it would make a lot of sense to me too.

**Comment by slingamn** [comment]
Currently on https://irc.darwin.network/ we are managing Oragono with systemd, having it log solely to stderr, and having the systemd journal capture the output. Here's the unit file:

````
[Unit]
Description=ircd
After=network.target

[Service]
Type=simple
User=oragono
WorkingDirectory=/home/oragono
ExecStart=/home/oragono/go/bin/oragono run --conf /home/oragono/ircd.yaml
ExecReload=/bin/kill -HUP $MAINPID
Restart=on-failure

[Install]
WantedBy=multi-user.target
````

Meanwhile, `ircd.yaml` logs a limited amount of information to stderr, which is then written to the systemd journal.

I think it's definitely nice to also support logging to a plain file, for the benefit of non-systemd platforms that have `logrotate` or the equivalent available.

The `ExecReload` command isn't ideal; as per the `systemd.service(5)` manpage, the `ExecReload` command should synchronously wait for the reload to complete. Back when the REST API was still present, this could have been implemented somewhat hackishly as `curl -d "" http://localhost:8090/rehash`, but I didn't think it was worth the trouble.

**Comment by enckse** [comment]
So, after not thinking about this for a long time (using logrotate for the file rotation all this time), I guess at this point I agree with @euank on the premise of that relying on systemd in this case is probably for the best. It's only my preference but maybe the "default" yaml file shouldn't set "file" as it's default method? Basically, if you leave that, you'll either have to realize you need to write a logrotate or end up with a really nice and big log file full of cool information but not useful when you need it!

Anyway, when the next release of oragono comes I'll probably update the aur packaging for arch slightly to switch that around a bit to push everything into stdout/stderr I'll probably also update the arch aur service from what @slingamn has (within reason), I honestly just wanted to start running oragono at the time and now I'm like "maybe this deserved some more effort". I agree that HUP is probably misaligned with intention of ExecReload but at the same time I don't know how much effort I would put into this either...

What I'm running now...
```
[Unit]
Description=A modern IRC server written in Go

[Service]
Type=simple
ExecStart=/usr/bin/oragono run --conf /etc/oragono.conf

[Install]
WantedBy=multi-user.target
```

tl;dr: logging to file is definitely still useful but maybe shouldn't be the default config setting ¯\\\_(ツ)\_/¯

**Comment by DanielOaks** [comment]
Yeah I'd be up for having a file method commented-out in the config file by default and just having it dump to stdout/srderr. Will probably make some of those changes in 0.12.0 along with the db update stuff we're pushing off 'til them.
